cirq/sim/act_on_args.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Objects and methods for acting efficiently on a state tensor."""
+import abc
+from typing import Any, Iterable, Dict, Sequence, List
+
+import numpy as np
+
+from cirq import protocols, ops
+from cirq.protocols.decompose_protocol import _try_decompose_into_operations_and_qubits
+
+
+class ActOnArgs:
+    """State and context for an operation acting on a state tensor."""
+
+    def __init__(
+        self,
+        axes: Iterable[int],
+        prng: np.random.RandomState,
+        log_of_measurement_results: Dict[str, Any],
+    ):
+        """
+        Args:
+            axes: The indices of axes corresponding to the qubits that the
+                operation is supposed to act upon.
+            prng: The pseudo random number generator to use for probabilistic
+                effects.
+            log_of_measurement_results: A mutable object that measurements are
+                being recorded into. Edit it easily by calling
+                `ActOnStateVectorArgs.record_measurement_result`.
+        """
+        self.axes = tuple(axes)
+        self.prng = prng
+        self.log_of_measurement_results = log_of_measurement_results
+
+    def measure(self, key, invert_mask):
+        """Adds a measurement result to the log.
+
+        Args:
+            key: The key the measurement result should be logged under. Note
+                that operations should only store results under keys they have
+                declared in a `_measurement_keys_` method.
+            invert_mask: The invert mask for the measurement.
+        """
+        bits = self.perform_measurement()
+        corrected = [bit ^ (bit < 2 and mask) for bit, mask in zip(bits, invert_mask)]
+        if key in self.log_of_measurement_results:
+            raise ValueError(f"Measurement already logged to key {key!r}")
+        self.log_of_measurement_results[key] = corrected
+
+    @abc.abstractmethod
+    def perform_measurement(self) -> List[int]:
+        """Child classes that perform measurements should implement this with
+        the implementation."""
+
+
+def strat_act_on_from_apply_decompose(
```

Comment:
This strategy can potentially modify the internal state halfway just to eventually fail when it encounters a decomposed op that isn't supported. This is a valid flow and we should be able to move on to the next strategy when the above happens. But having modified the state partially leads to a bug and wrong outputs.

Add a way to restore the internal state to the starting value when this happens. #3874 has that implemented but I will be closing that PR in favor of this one. Feel free to pick up that logic.

Reply:
I'd like to do that in a separate PR if possible. For now the decompose is always the last operation called, so failure would just mean the circuit can't be simulated. The problem with creating a checkpoint is that the checkpoints can be multi-GB in size, so that would be a big perf regression.

I'll look at the on_stuck thing that Balint mentioned and see if restoring that code helps here.

Reply:
One option would be, instead of having each ActOnXXXArgs implement `_act_on_fallback_`, we could implement it in the ActOnArgs base class. The implementation would look like

```
self._act_on_fallback_impl
if allow_decompose:
  _strat_decompose(self, action)
```

such that `_act_on_fallback_impl` is an abstract method implemented in the child classes, and now _strat_decompose is private so we know it should always be the last strategy called before giving up, so no need to copy the whole state tensor each time it executes.

I think this change could go in a separate PR though, since currently it's not required.